### PR TITLE
The tx-seam prohibition follows a call one hop further

### LINKED ISSUE #4118: The nested-acquire gate reads one call deep, so a wrapper hides a pool deadlock
## What is wrong

`TestATxAcceptingFunctionAcquiresNoConnectionOfItsOwn` (`backend/gates/txseamacquire_test.go`) matches calls inside a tx-borrowing body against `connectionAcquirers`, a hand-maintained map of names. It does not follow the bodies of the functions it sees, so an acquirer reached one hop away is invisible.

## Failure scenario

```go
// inside WithWorkspaceTx(ctx, pool, func(tx pgx.Tx) error { … })
decorateRows(ctx, e.names, rows)   // gate sees an unregistered name, passes
```

```go
func decorateRows(ctx context.Context, names attention.Names, rows []map[string]any) {
    labelDerivationRows(ctx, names, "deal", rows)   // registered, but never reached by the walk
}
```

The wrapper takes no `pgx.Tx`, so it is not a tx-borrowing body of its own, and the walk never descends into it. The deadlock is exactly as real as the one the gate exists to catch: each label read takes a second connection while the outer transaction holds the first, and with the pool at 16 a handful of concurrent requests wedge the whole API.

## Why this is not already fixed

The gate was widened (in the drill-through PR) to read bare-identifier calls as well as method calls, which closed the direct case — a registered acquirer called by name inside a borrowed transaction now fails. One hop of indirection still passes.

Closing that means resolving package-local calls and walking their bodies transitively, with cycle handling and a decision about how far to follow across files. That is a different gate from the one that exists, and worth doing deliberately rather than as a side effect.

## Related

The registry approach has a second edge: `connectionAcquirers` is hand-maintained, so a NEW acquirer nobody registers is invisible however it is called. Deriving the acquirer set from the code — functions whose bodies reach `pool.Acquire`/`BeginTx` — would close both edges at once and is probably the same piece of work.

Found by an adversarial review of the drill-through PR.

### PR DESCRIPTION
**Pull request by dungbanguyen**
Closes #4118.

The gate matched the calls it could **see** against a list of acquirer names. A wrapper takes no `pgx.Tx`, so it is not a borrowing body of its own and nothing judged it:

```go
// inside WithWorkspaceTx(ctx, pool, func(tx pgx.Tx) error { … })
decorateRows(ctx, e.names, rows)                     // unregistered name: passed

func decorateRows(...) { labelDerivationRows(...) }  // registered, never reached
```

The deadlock is identical either way: a second connection taken while the outer transaction holds the first, undetectable by Postgres, and a pool of 16 wedges the API under a handful of concurrent requests.

The gate now asks **"does any chain from here reach an acquirer"** rather than "is this call one". That closes the ticket's second edge for free — an acquirer nobody added to `connectionAcquirers` is found anyway, as long as it is reached from the package it was written in.

## The part worth reviewing: two wrong drafts, both measured

I did not reason my way to the scope. I ran it.

**Draft 1 — follow selector calls by method name.** Seventeen findings, **every one false**. `privacy` holds both `e.blob.Delete` (an object store, touches no pool) and `PolicyStore.Delete` (opens a transaction). A name cannot tell them apart.

**Draft 2 — index methods too.** Three more, because `StageSemantic(*in.Semantic)` is a **type conversion** that shares its name with `(*Store).StageSemantic`. A `switch` statement was reported as a deadlock.

A gate that cries wolf gets deleted by the next person who sees it red, so the line is the one Go itself draws: a bare identifier resolves package-locally with no type information, and the index holds **package-level functions only**, because that is all such a call can reach. Resolving a receiver needs full type checking, and a gate that guessed would name the wrong file with total confidence.

Both drafts are in the file's comments, and both are pinned as tests — `TestTheReachWalkStopsAtAReceiverCall` and `TestTheReachWalkIsNotFooledByAConversionSharingAMethodName` — so the next person to think "why not just follow selectors" gets the answer from a failing test rather than rediscovering seventeen false positives.

## The limits are tested, not asserted

Two blind spots remain and both are narrower than before: a call through a **receiver**, and a call **across a package edge**. `TestTheReachWalkStopsAtAReceiverCall` and `TestTheReachWalkStopsAtThePackageEdge` pin them, so they are facts rather than comments someone hopes are current.

## Fail-short guard

The walk reports only when it **follows** something, so an index reading a smaller tree finds nothing and reports PASS having checked one hop of nothing. The index count is asserted: **8574** functions today against a floor of **2000**.

## Checks

- `make check-backend` green; `./gates` green.
- **Mutation-checked** three ways: the walk following nothing fails both chain cases; the floor raised above the real count fails with the count; and the reported wrapper shape fails without the fix.
- Gate inventory regenerated with `-update-gate-inventory` rather than hand-edited.

## Note

`connectionAcquirers` still lists the roots — `Acquire`, `Begin`, `Tx`, `WithWorkspaceTx` and the catalog reads. That is deliberate: "this call takes a connection from the pool" is not something this tree can prove about a library primitive. The entries above the primitives are now reachable by derivation where they are package-local.

## Summary by CodeRabbit

- **Improvements**
  - Transaction-safety checks now trace indirect connection acquisition through multi-step call chains.
  - Improved accuracy across package boundaries, cycles, conversions, receiver-based calls, parameters, and local values.
  - Direct connection acquisitions continue to be reported in source order, while indirect paths include their acquisition chain.

- **Documentation**
  - Updated the gate inventory to include the expanded transaction-reachability coverage.

- **Tests**
  - Added coverage for chained calls, cycles, indirect reachability, and boundary conditions.

### CODE REVIEW

File: backend/gates/txseamreach_test.go [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+		if lit, ok := n.(*ast.FuncLit); ok {
+			// A tx-taking literal is judged as a body of its own, exactly as
+			// acquires() leaves it alone.
+			return !takesPgxTx(lit.Type.Params, b.pgxName)
+		}
+		call, ok := n.(*ast.CallExpr)
+		if !ok {
+			return true
+		}
+		if fn, ok := call.Fun.(*ast.Ident); ok {
```
[the remark this answers is not in the document]
Comment (dungbanguyen) [reply backend/gates/txseamreach_test.go]: Fixed. This is the **third** false-positive class in this walk — after selector calls and type conversions — and it is the same mistake each time: a name is not a reference. Given the argument I made in the PR description for why this gate is worth having, that is exactly the one I cannot afford to leave in.

`calledNames` now excludes names the body binds itself: parameters, results, `:=` targets, `var`/`const` specs, range variables and nested literal parameters.

**One deliberate choice worth flagging.** I over-collect: a name bound *anywhere* in the body counts as bound throughout it, so a real call whose name a local shadows *later* is missed. That is a miss rather than a false accusation, and it is the right direction for this gate — proper block scoping needs the type information the walk deliberately does without, and a walk that guessed at scope would be back to naming the wrong file with confidence.

**Three fixtures, not two.** A parameter, a local function value, and — as the control — an unbound name that must still be followed. Without the third, the first two are equally satisfied by a walk that follows nothing at all, which is the way this fix would silently turn the gate off.

**On the mutation check, because it caught me out:** my first attempt deleted the `!bound[...]` guard, which left `bound` unused, so the package did not build — and a test binary that never runs reports no failures. It looked like the tests passed without the fix. The compiling form returns an empty set from `locallyBound`; under it both shadowing cases fail and the control still passes.

**Comment by dungbanguyen** [comment]
Pushed a fix for the red CI, and the cause is worth recording because local was green throughout.

**#4116 landed while this was open and added `labelDerivationRows` to `connectionAcquirers`** — the exact name my fixture borrowed from the reported defect. So the walk found it as a *direct* acquirer one hop earlier and returned a two-step chain, correctly, against a test asserting three:

```
the walk reached an acquirer from "decorateRows" via [decorateRows labelDerivationRows],
want [decorateRows labelDerivationRows Acquire]
```

The fixture had been testing the REACH and quietly started testing the direct match. Invented names now — a fixture that borrows a real acquirer's name is coupled to a registry it has no reason to track, and fails the day somebody extends it.

I checked the other fixtures for the same coupling; none of their names appears in `connectionAcquirers`.

`main` is merged in and `make check-backend` is green.
